# 404 Candidate templates

This notebook is to train on a dataset of 404 candidate templates. These include normalised tokens like `first_nfkd_0` to see if we can predict normalisation as well as email templates.

In [1]:
!pip install lightgbm
!pip install sqlalchemy

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 42.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.3/3.3 MB 39.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 585.5/585.5 kB 57.5 MB/s eta 0:00:00


In [2]:
import json
import re
import random
import matplotlib.pyplot as plt
import lightgbm as lgb
import pandas as pd
import numpy as np
import sqlite3
from datetime import datetime
from collections import defaultdict
from sqlalchemy import select, func
from sklearn.metrics import precision_score, recall_score, f1_score

In [3]:
from google.colab import drive

drive.mount("/content/drive")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
# Paths to files in Drive
db_path = "/content/drive/MyDrive/Colab Notebooks/database.db"
train_ids_path = "/content/train_ids.csv"
val_ids_path = "/content/validation_ids.csv"

Pull whole dataset from SQL db into pandas frame.

In [5]:
conn = sqlite3.connect(db_path)  # or your local path
full_df = pd.read_sql_query("SELECT * FROM feature_matrix", conn)

Saved the training and validation IDs in a CSV so we can split the data set.

In [6]:
train_ids = pd.read_csv(train_ids_path)["train_ids"].dropna().astype(int).tolist()
val_ids = pd.read_csv(val_ids_path)["validation_ids"].dropna().astype(int).tolist()

Validation dataset pulled from full set, remianing is training set. This is to ensure that we don't train on validation set.

In [7]:
val_ids_set = set(val_ids)
val_df = full_df[full_df["clean_row_id"].isin(val_ids_set)]
full_df = full_df[~full_df["clean_row_id"].isin(val_ids_set)]

This was pulled from the initial model development notebook.

In [8]:
def compute_ranking_metrics(df, k=3):
    # Group
    grouped = df.groupby("clean_row_id")

    # Get top 1 and calculate accuracy
    top1 = grouped.apply(lambda g: g.loc[g["score"].idxmax()]).reset_index(drop=True)
    acc1 = (top1["label"] == 1).mean()

    # Same with recall
    topk = grouped.apply(lambda g: g.nlargest(k, "score")).reset_index(drop=True)
    recall_k = topk.groupby("clean_row_id")["label"].max().mean()

    # MMR
    def reciprocal_rank(g):
        sorted_g = g.sort_values("score", ascending=False).reset_index()
        match = sorted_g[sorted_g["label"] == 1]
        return 1.0 / (match.index[0] + 1) if not match.empty else 0.0

    mrr = grouped.apply(reciprocal_rank).mean()
    return acc1, recall_k, mrr

This is so we can save logs to file in case collab logs us out and we lose our printouts.

In [9]:
# Paths
timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
log_path = f"/content/drive/MyDrive/lgb_log_{timestamp}.txt"
model_path = f"/content/drive/MyDrive/lgb_model_{timestamp}.txt"

# Logger
log_file = open(log_path, "w")


def log_callback(period=10):
    def _callback(env):
        if env.iteration % period == 0 or env.iteration + 1 == env.end_iteration:
            result = f"[{env.iteration}] "
            for data_name, eval_name, result_val, _ in env.evaluation_result_list:
                result += f"{data_name} {eval_name}: {result_val:.5f}  "
            print(result)
            log_file.write(result + "\n")
            log_file.flush()

    return _callback

From the model development notebook, changed slightly to fit our dataframes.

In [10]:
def train_model(
    train_ids,
    val_ids,
    parameters: dict,
    n_rounds: int = 500,
    lr_decay_gamma: float = 0.95,
):
    conn = sqlite3.connect(db_path)
    # Load validation set
    X_val = val_df.drop(
        columns=["label", "clean_row_id", "investor", "firm", "template_id"]
    )
    y_val = val_df["label"]
    val_group_sizes = val_df.groupby("clean_row_id").size().tolist()
    lgb_val = lgb.Dataset(
        X_val, label=y_val, group=val_group_sizes, free_raw_data=False
    )

    # Load full training set
    full_df
    X_train = full_df.drop(
        columns=["label", "clean_row_id", "investor", "firm", "template_id"]
    )
    y_train = full_df["label"]
    train_group_sizes = full_df.groupby("clean_row_id").size().tolist()
    lgb_train = lgb.Dataset(
        X_train, label=y_train, group=train_group_sizes, free_raw_data=False
    )

    # Learning rate schedule
    def lr_decay(current_round):
        return parameters["learning_rate"] * (lr_decay_gamma**current_round)

    # Train model
    model = lgb.train(
        params=parameters,
        train_set=lgb_train,
        num_boost_round=n_rounds,
        valid_sets=[lgb_train, lgb_val],
        valid_names=["train", "val"],
        callbacks=[
            lgb.reset_parameter(learning_rate=lr_decay),
            lgb.early_stopping(stopping_rounds=500),
            lgb.log_evaluation(period=1),
            log_callback(1),
        ],
    )

    # Predict and Evaluate
    preds = model.predict(X_val, num_iteration=model.best_iteration)
    val_df["score"] = preds

    acc1, recall3, mrr = compute_ranking_metrics(val_df, k=3)

    print("\Evaluation Metrics (Validation Set):")
    print(f"Accuracy@1 : {acc1:.4f}")
    print(f"Recall@3   : {recall3:.4f}")
    print(f"MRR        : {mrr:.4f}")

    return model

In [11]:
params = {
    "objective": "lambdarank",
    "metric": ["ndcg"],
    "eval_at": [1, 3],
    "label_gain": [0, 1],
    "learning_rate": 0.1,
    "num_leaves": 128,
    "max_depth": 10,
    "min_split_gain": 1e-3,
    "min_child_weight": 0.01,
    "max_delta_step": 1.0,
    "scale_pos_weight": 403,
    "verbosity": -1,
    "boosting_type": "gbdt",
    "force_row_wise": True,
    "max_bin": 128,
    "feature_fraction": 0.8,
    "bagging_freq": 0.8,
    "bagging_freq": 1,
    "lambda_l1": 1.0,
    "min_data_per_group": 10,
}

In [12]:
# Retrain
model = train_model(train_ids, val_ids, params, n_rounds=2000)
log_file.close()
model.save_model(model_path)

[0] train ndcg@1: 0.87845  train ndcg@3: 0.94357  val ndcg@1: 0.74552  val ndcg@3: 0.86608  
[1]	train's ndcg@1: 0.878452	train's ndcg@3: 0.943567	val's ndcg@1: 0.745519	val's ndcg@3: 0.866079
Training until validation scores don't improve for 500 rounds
[1] train ndcg@1: 0.87845  train ndcg@3: 0.94357  val ndcg@1: 0.74552  val ndcg@3: 0.86608  
[2]	train's ndcg@1: 0.878452	train's ndcg@3: 0.943567	val's ndcg@1: 0.745519	val's ndcg@3: 0.866079
[2] train ndcg@1: 0.88788  train ndcg@3: 0.95227  val ndcg@1: 0.78110  val ndcg@3: 0.89816  
[3]	train's ndcg@1: 0.887884	train's ndcg@3: 0.952266	val's ndcg@1: 0.7811	val's ndcg@3: 0.898162
[3] train ndcg@1: 0.89214  train ndcg@3: 0.95431  val ndcg@1: 0.79470  val ndcg@3: 0.90501  
[4]	train's ndcg@1: 0.892141	train's ndcg@3: 0.954306	val's ndcg@1: 0.794697	val's ndcg@3: 0.905006
[4] train ndcg@1: 0.89462  train ndcg@3: 0.95636  val ndcg@1: 0.80349  val ndcg@3: 0.91204  
[5]	train's ndcg@1: 0.894622	train's ndcg@3: 0.956358	val's ndcg@1: 0.80349

/tmp/ipython-input-8-2860974315.py:6: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  top1 = grouped.apply(lambda g: g.loc[g["score"].idxmax()]).reset_index(drop=True)
/tmp/ipython-input-8-2860974315.py:10: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  topk = grouped.apply(lambda g: g.nlargest(k, "score")).reset_index(drop=True)


\Evaluation Metrics (Validation Set):
Accuracy@1 : 0.8542
Recall@3   : 0.9913
MRR        : 0.9216


Same as before, good sign.

NDCG@1: 0.854224 - precision at top 1 spot is correct almost 85% of the time.

NDCG@3: 0.938389 - at top3, it is correct almost 93% of the time.

Accuracy@1: 0.8542, Recall@3: 0.9913 - indicating the same thing, most of the time our top three ranking predictions contains the correct template.